Cancel Runs on ClearAllObjects

kateforsberg13 opened the pull request:
closes #502 
Update run cancelling to fully reset all of the nodes and the state machines. 
Adds new reset_machine and reset methods in the context objects and in the machines.

Removed cancellation when trying to run a flow 2x back to back. Now it just prevents that behavior from happening, and the flow continues to run. 

There is some weirdness with outputs still - but that will have to do with being able to cancel the threads, which is not handled in this PR.

SavagePencil commented on src/griptape_nodes/exe_types/flow.py:
```diff
@@ -191,13 +191,13 @@ def cancel_flow_run(self) -> None:
-        # Cancel all future runs
         del self.flow_queue
         self.flow_queue = Queue()
+        if self.control_flow_machine.get_current_state():
+            self.control_flow_machine.reset_machine()
+        else:
+            self.control_flow_machine._context.resolution_machine.reset_machine()
```
what puts us in a situation where `get_current_state` is returning None, and why do we have to go up to the private var?

kateforsberg13 replied:
When we're doing single node resolution, it only starts the resolution_machine. but i guess we dont need to check state, we can just cancel.

SavagePencil commented on src/griptape_nodes/retained_mode/griptape_nodes.py:
```diff
@@ -497,6 +497,18 @@ def on_clear_all_object_state_request(self, request: ClearAllObjectStateRequest)
             return ClearAllObjectStateResultFailure()
         # Let's try and clear it all.
         context_mgr = GriptapeNodes.ContextManager()
+        if context_mgr.has_current_flow():
```
I don't think we need to be checking the context manager here. Running Flows may not be 1:1 with things on the CurrentContext stack.

kateforsberg13 replied:
oh ok. I thought that was what the context manager kept track of. like which flow was being interacted with

kateforsberg13 commented:
This still doesn't totally solve the problem. Especially for streaming. We still need to solve the threading events. but this is a good start.

SavagePencil commented on src/griptape_nodes/retained_mode/griptape_nodes.py:
```diff
@@ -496,6 +496,14 @@ def on_clear_all_object_state_request(self, request: ClearAllObjectStateRequest)
+            if flow.check_for_existing_running_flow():
+                result = GriptapeNodes.handle_request(CancelFlowRequest(flow_name=flow_name))
+                if not result.succeeded():
+                    details = "Attempted to clear all object state and delete everything. Failed because running flow could not cancel."
```
Can we include the flow name?

kateforsberg13 replied:
yep

SavagePencil commented on src/griptape_nodes/retained_mode/griptape_nodes.py:
```diff
@@ -691,7 +699,7 @@ def does_canvas_exist(self) -> bool:
 
     def on_create_flow_request(self, request: CreateFlowRequest) -> ResultPayload:
-        obj_mgr = GriptapeNodes().get_instance().ObjectManager()
+        obj_mgr = GriptapeNodes().get_instance()._object_manager
```
instead of referencing private variables, use `GriptapeNodes.ObjectManager()`

kateforsberg13 replied:
yep. sorry must have missed just this one

collindutter replied:
Seems this got prematurely resolved

SavagePencil commented on src/griptape_nodes/retained_mode/griptape_nodes.py:
```diff
@@ -2169,7 +2173,7 @@ def on_remove_parameter_from_node_request(self, request: RemoveParameterFromNode
 
         # No tricky stuff, users!
-        if parameter.user_defined is False:
+        if parameter and parameter.user_defined is False:
```
we already established parameter as valid, no?

kateforsberg13 replied:
it was giving me a ruff error that is wasn't valid for some reason.

kateforsberg13 replied:
```
if parameter is None and parameter_group is None:
            details = f"Attempted to remove Parameter '{request.parameter_name}' from Node '{node_name}'. Failed because it didn't have a Parameter with that name on it."
            logger.error(details)

            result = RemoveParameterFromNodeResultFailure()
            return result
        if parameter_group is not None:
            for child in parameter_group.find_elements_by_type(Parameter):
                GriptapeNodes.handle_request(RemoveParameterFromNodeRequest(child.name, node_name))
            node.remove_group_by_name(request.parameter_name)
            return RemoveParameterFromNodeResultSuccess()

        # No tricky stuff, users!
        if parameter and parameter.user_defined is False:
```

see technically, we know parameter is valid. but i don't know if the code editor is smart enough for this. which is why I got the error.

collindutter replied:
I think Pyright is correct here. Based on our current logic, `parameter: None`, `parameter_group: value` would slip by.

SavagePencil commented on src/griptape_nodes/retained_mode/griptape_nodes.py:
```diff
@@ -2406,8 +2416,12 @@ def on_alter_parameter_details_request(self, request: AlterParameterDetailsReque
         # Does this node exist?
-        obj_mgr = GriptapeNodes().get_instance().ObjectManager()
+        obj_mgr = GriptapeNodes.ObjectManager()
+        if not node_name:
```
have we not guaranteed `node_name` as valid here?

kateforsberg13 replied:
Ruff was leaving an error that I can't pass "str|None" in that function. Somehow updating ObjectManager() is creating this, I think that somehow it must not be guaranteed (or the checking isn't picking up on it)

collindutter replied:
I think making that update _fixed_ type issues that were previously being suppressed. Pyright is in the right here.

SavagePencil commented on src/griptape_nodes/retained_mode/griptape_nodes.py:
```diff
@@ -1976,8 +1977,11 @@ def on_list_parameters_on_node_request(self, request: ListParametersOnNodeReques
-        obj_mgr = GriptapeNodes().get_instance().ObjectManager()
-
+        obj_mgr = GriptapeNodes.ObjectManager()
+        if not node_name:
```
have we not guaranteed `node_name` as valid here?

kateforsberg13 replied:
same issue as above

kateforsberg13 replied:
<img width="564" alt="Screenshot 2025-04-23 at 4 18 01 PM" src="https://github.com/user-attachments/assets/4bbf1f2a-9bd5-4fdd-b5b8-e85ea38ccd17" />
before it looked like this: so because it was getting the object manager like that, it wasn't doing the checking. 
<img width="1402" alt="Screenshot 2025-04-23 at 4 18 40 PM" src="https://github.com/user-attachments/assets/c04828eb-bb1a-4fa1-a955-988f9b15cf21" />
Now it's actually grabbing the method, and throwing this type checking.

collindutter replied:
This is valid pyright warning, `node_name` _could_ be `None` here.

SavagePencil commented on src/griptape_nodes/retained_mode/griptape_nodes.py:
```diff
@@ -2463,12 +2477,17 @@ def on_get_parameter_value_request(self, request: GetParameterValueRequest) -> R
         param_name = request.parameter_name
 
         # Get the node
+        if not node_name:
+            details = f"Attempted to get value for Parameter '{request.parameter_name}' from node in the Current Context. Failed because no node name was provided."
+            logger.error(details)
+
+            return GetParameterValueResultFailure()
```
should be guaranteed?

kateforsberg13 replied:
same issue as above. ruff errors

collindutter commented on src/griptape_nodes/machines/node_resolution.py:
```diff
@@ -50,6 +50,21 @@ def __init__(self, flow: ControlFlow) -> None:
         self.paused = False
 
+    def reset(self) -> None:
+        if len(self.focus_stack) > 0:
```
Nit: can be a bit more pythonic with `if self.focus_stack`

collindutter commented on src/griptape_nodes/retained_mode/griptape_nodes.py:
```diff
@@ -2169,7 +2173,7 @@ def on_remove_parameter_from_node_request(self, request: RemoveParameterFromNode
 
         # No tricky stuff, users!
-        if parameter.user_defined is False:
+        if parameter and parameter.user_defined is False:
```
Use explicit `if parameter is not None` check instead of `if parameter`

collindutter commented on src/griptape_nodes/retained_mode/griptape_nodes.py:
```diff
@@ -1976,8 +1977,11 @@ def on_list_parameters_on_node_request(self, request: ListParametersOnNodeReques
-        obj_mgr = GriptapeNodes().get_instance().ObjectManager()
-
+        obj_mgr = GriptapeNodes.ObjectManager()
+        if not node_name:
```
Use explicit `if node_name is not None` check.

collindutter commented on src/griptape_nodes/retained_mode/griptape_nodes.py:
```diff
@@ -2406,8 +2416,12 @@ def on_alter_parameter_details_request(self, request: AlterParameterDetailsReque
         # Does this node exist?
-        obj_mgr = GriptapeNodes().get_instance().ObjectManager()
+        obj_mgr = GriptapeNodes.ObjectManager()
+        if not node_name:
```
Explicit `is not None`.

collindutter commented on src/griptape_nodes/retained_mode/griptape_nodes.py:
```diff
@@ -2463,12 +2477,17 @@ def on_get_parameter_value_request(self, request: GetParameterValueRequest) -> R
         param_name = request.parameter_name
 
         # Get the node
+        if not node_name:
```
Explicit `is not None`.

kateforsberg13 commented:
I just went through and removed all `GriptapeNodes()` instances in `griptape_nodes.py`, which exposed more pyright errors that I fixed. I also got rid of all the cases I could find without explicit `is None` or `is not None`, since this is what keeps getting us on the `ParameterList` stuff.

SavagePencil commented on src/griptape_nodes/retained_mode/griptape_nodes.py:
```diff
@@ -1976,8 +1982,11 @@ def on_list_parameters_on_node_request(self, request: ListParametersOnNodeReques
-        obj_mgr = GriptapeNodes().get_instance().ObjectManager()
-
+        obj_mgr = GriptapeNodes.ObjectManager()
+        if node_name is None:
```
I'm still not seeing how `node_name` can be None here. It's assigned to the `request.node_name`. If that is None, it checks to see if the ContextManager has one. If it doesn't, we abort. If it DOES have one (it's a str), we assign it to `node_name`. How do we get through here with a None?

kateforsberg13 replied:
```
node_name = request.node_name

        if request.node_name is None:
            # Get from the current context.
            if not GriptapeNodes.ContextManager().has_current_node():
                details = "Attempted to list Parameters for a Node from the Current Context. Failed because the Current Context is empty."
                logger.error(details)
                return ListParametersOnNodeResultFailure()
```

because it's checking is `request.node_name` is `None`, not `node_name`. I changed it to `node_name` and it removes the pyright error.
